pkg/kubefed2/unjoin.go

```diff
@@ -0,0 +1,409 @@
+func deleteRBACResources(hostClusterClientset, unjoiningClusterClientset client.Interface,
+	namespace, unjoiningClusterName, hostClusterName string, dryRun bool) error {
+
+	saName := util.ClusterServiceAccountName(unjoiningClusterName, hostClusterName)
+
+	glog.V(2).Infof("Deleting cluster role binding for service account: %s in unjoining cluster: %s",
+		saName, unjoiningClusterName)
+
+	err := deleteClusterRoleAndBinding(unjoiningClusterClientset, saName, namespace,
+		unjoiningClusterName, dryRun)
+	if err != nil {
+		glog.V(2).Infof("Error deleting cluster role binding for service account: %s in unjoining cluster: %v",
+			saName, err)
+		return err
+	}
+
+	glog.V(2).Infof("Deleted cluster role binding for service account: %s in unjoining cluster: %s",
+		saName, unjoiningClusterName)
+
+	glog.V(2).Infof("Deleting service account %s in unjoining cluster: %s", saName, unjoiningClusterName)
+
+	err = deleteServiceAccount(unjoiningClusterClientset, saName, namespace, dryRun)
+	if err != nil {
+		return err
+	}
+
+	glog.V(2).Infof("Deleted service account %s in unjoining cluster: %s", saName, unjoiningClusterName)
+
+	return nil
+}
+
+// deleteFedNSFromUnjoinCluster deletes the federation namespace and multi cluster
+// public namespace in the unjoining cluster.
+func deleteFedNSFromUnjoinCluster(unjoiningClusterClientset client.Interface,
+	federationNamespace, unjoiningClusterName, hostClusterName string, dryRun bool) error {
+
+	if dryRun {
+		return nil
+	}
+
+	err := unjoiningClusterClientset.CoreV1().Namespaces().Delete(federationNamespace,
+		&metav1.DeleteOptions{})
+	if err != nil {
+		glog.V(2).Infof("Could not delete federation namespace: %s due to: %v", federationNamespace, err)
+		return err
+	}
+
+	return nil
+}
+
+// deleteServiceAccount deletes a service account in the cluster associated
+// with clusterClientset with credentials that are used by the host cluster
+// to access its API server.
+func deleteServiceAccount(clusterClientset client.Interface, saName,
+	namespace string, dryRun bool) error {
+	if dryRun {
+		return nil
+	}
+
+	// Delete a service account.
+	err := clusterClientset.CoreV1().ServiceAccounts(namespace).Delete(saName,
+		&metav1.DeleteOptions{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// deleteClusterRoleAndBinding deletes an RBAC cluster role and binding that
+// allows the service account identified by saName to access all resources in
+// all namespaces in the cluster associated with clusterClientset.
+func deleteClusterRoleAndBinding(clusterClientset client.Interface, saName, namespace,
+	unjoiningClusterName string, dryRun bool) error {
+
+	if dryRun {
+		return nil
+	}
+
+	roleName := util.ClusterRoleName(saName)
+
+	err := clusterClientset.RbacV1().ClusterRoleBindings().Delete(roleName, &metav1.DeleteOptions{})
+	if err != nil {
+		glog.V(2).Infof("Could not delete cluster role binding for service account: %s in unjoining cluster: %v",
+			saName, err)
+		return err
```

Comment:
Why does it make sense to log extensively but not annotate the error with the same details?  Troubleshooting an error then requires manually correlating an error with the log, vs having sufficient detail in the log.

Reply:
So you mean we can make the log simple, such as follows, as the `err` should already include enough detail?

```
glog.V(2).Infof("Could not delete cluster role binding: %v", err)
```
If so, how about do a swipe for both join and unjoin, and handle this in a follow up PR?